staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"spec": *objectTypePtr(map[string]schema.Structural{
+							"field1": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.spec.field1 == 'a'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata)": "undefined field 'metadata'",
+			},
+		},
+		{name: "embedded with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			errors: map[string]string{
+				// we only provide information in the schema to validation rules
+				// without type=object, we don't provide access to the embedded resource
+				"has(self.embedded)": "undefined field 'embedded'",
+			},
+		},
+		{name: "embedded object with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			valid: []string{
+				"has(self.embedded)",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata)": "undefined field 'metadata'",
+				"has(embedded.spec)":     "undefined field 'spec'",
+			},
+		},
+		{name: "string in intOrString",
+			obj: map[string]interface{}{
+				"something": map[string]interface{}{
+					"strVal": "25%",
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"something": intOrStringType(),
+			}),
+			valid: []string{"has(something.strVal) && something.strVal == '25%'"},
```

Comment:
add the intVal variant too. It should type correctly.

Reply:
Added above. Also, per discussion with @liggitt we're switching to using a Dynamic type for int-or-string, so instead of `has(something.strVal)...` we do `type(something) == string && something == '25%'`